Bearded.Utilities/Monads/Result.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Bearded.Utilities.Monads
+{
+    public readonly struct Result<TResult, TError> : IEquatable<Result<TResult, TError>>
+    {
+        private readonly bool isSuccess;
+        private readonly TResult result;
+        private readonly TError error;
+
+        private Result(bool isSuccess, TResult result, TError error)
+        {
+            this.isSuccess = isSuccess;
+            this.result = result;
+            this.error = error;
+        }
+
+        public TResult ResultOrDefault(TResult @default) => isSuccess ? result : @default;
+
+        public TResult ResultOrDefault(Func<TResult> defaultProvider) => isSuccess ? result : defaultProvider();
+
+        public TResult ResultOrThrow(Func<TError, Exception> exceptionProvider) =>
+            isSuccess ? result : throw exceptionProvider(error);
+
+        public Result<TOut, TError> Select<TOut>(Func<TResult, TOut> selector) =>
+            isSuccess ? (Result<TOut, TError>) Result.Success(selector(result)) : Result.Failure(error);
+
+        public Result<TOut, TError> SelectMany<TOut>(Func<TResult, Result<TOut, TError>> selector) =>
+            isSuccess ? selector(result) : Result.Failure(error);
+
+        public void Match(Action<TResult> onSuccess)
+        {
+            if (isSuccess)
+            {
+                onSuccess(result);
+            }
+        }
+
+        public void Match(Action<TResult> onSuccess, Action<TError> onFailure)
+        {
+            if (isSuccess)
+            {
+                onSuccess(result);
+            }
+            else
+            {
+                onFailure(error);
+            }
+        }
+
+        public TOut Match<TOut>(Func<TResult, TOut> onSuccess, Func<TError, TOut> onFailure) =>
+            isSuccess ? onSuccess(result) : onFailure(error);
+
+        public bool Equals(Result<TResult, TError> other) =>
+            isSuccess == other.isSuccess
+            && EqualityComparer<TResult>.Default.Equals(result, other.result)
+            && EqualityComparer<TError>.Default.Equals(error, other.error);
+
+        public override bool Equals(object obj) => obj is Result<TResult, TError> other && Equals(other);
+
+        public static bool operator ==(Result<TResult, TError> left, Result<TResult, TError> right) =>
+            left.Equals(right);
+
+        public static bool operator !=(Result<TResult, TError> left, Result<TResult, TError> right) =>
+            !left.Equals(right);
+
+        public override int GetHashCode()
+        {
+            var hashCode = isSuccess.GetHashCode();
+            hashCode = (hashCode * 397) ^ EqualityComparer<TResult>.Default.GetHashCode(result);
+            hashCode = (hashCode * 397) ^ EqualityComparer<TError>.Default.GetHashCode(error);
+            return hashCode;
```

Comment:
Should we start using the .NET provided hashcode functionality?

Reply:
I can't access `System.HashCode` right now. My guess is that is because we're still targeting netframework at `developer` for this PR.

Happy to do a pass over all our code to switch to the .NET provided functionality in one go after our official switchover.

Reply:
Updated now that we have .NET Core